### PR TITLE
Pt1 Remove legacy request/response classes

### PR DESCRIPTION
**Pull request by dagnir**
Flesh out SdkRequest and SdkResponse classes and update code generation
to emit service request/responses using these classes.

This only contains the bare minimum `core` and `codegen` changes, so it doesn't build properly at the moment.

See the test files like `codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/alltypesrequest.java` and 
`codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/alltypesresponse.java`

for examples of how the generated files look.

### CODE REVIEW

**Comment by dagnir** [comment]
Related to #33

**Comment by millems** [comment]
@shorea : This may be what you're suggesting, but I was thinking we have an SdkRequest/Response base class, an AwsRequest/Response base class (extends Sdk classes) and an ApiGatewayRequest/Response base class (extends Sdk classes). They'd each contain the feature specific to their slice of required functionality. 

With that, we could have ClientHandler extensions per-SDK type that takes the per-SDK type request, extracts the SDK type-specific features and loads it into the execution attributes to be used by its interceptors and signers. 

Here's some hand-wavy example code where we have removed `AwsCredentialsProvider` from core:
```java
class AwsRequest extends SdkRequest {
   public AwsCredentialsProvider credentialsProvider();
}

class AwsSyncClientConfiguration extends SdkSyncClientConfiguration {
    // Note: SdkSyncClientConfiguration is the current SyncClientConfiguration, renamed for clarity.
    public AwsCredentialsProvider credentialsProvider();
}

class AwsClientHandler extends SdkClientHandler<AwsRequest> {
    private final AwsSyncClientConfiguration clientConfiguration;

    /**
     * Called by the service client when it is created.
     */
    public AwsClientHandler(AwsSyncClientConfiguration clientConfiguration) {
        super(clientConfiguration);
        this.clientConfiguration = clientConfiguration;
    }
    
    /**
     * Called by SdkClientHandler to attached aws-specific attributes from the request to the execution. 
     * These attributes are then used in AWS-specific interceptors and AWS-specific signers.
     */
    @Override
    public extractAttributes(AwsRequest request, ExecutionAttributes attributes) {
        attributes.set(AwsExecutionAttributes.CREDENTIALS, 
                       request.credentialsProvider().orElse(clientConfiguration.credentialsProvider()).getCredentials());
    }
}
```

If we wanted to hide features in certain client types (eg. if there's an SDK feature we want to hide in the API gateway clients), we could do it via more limited interfaces on the service clients.

This approach makes it a bit tricky to share functionality between certain client types (eg. if ApiG and AWS Requests need to share a feature, but not have it in SDK Request), but covering for that feels like we might be trying a bit too hard. We could always solve the presentation of that to our customers via shared interfaces and implement it under the hood using delegation in both the ApiG and AWS base classes to a shared library.

+1 To the idea of service-specific base model classes. Other than just for instanceof help, I don't know, maybe someone would want to disable S3 accelerate per-request?.

File: core/src/main/java/software/amazon/awssdk/SdkRequest.java [hunk 1]
```diff
@@ -15,12 +15,94 @@
...
  * The base class for all SDK requests.
  *
- * TODO: SDK-specific options on the {@link AmazonWebServiceRequest} should be migrated here as part of the base-model refactor.
- *
  * @see SdkResponse
  */
-public abstract class SdkRequest {
+public abstract class SdkRequest<B extends SdkRequest.Builder<B, R>, R extends SdkRequest<B, R>>
+        implements ToCopyableBuilder<B, R> {
+    private final RequestClientOptions requestClientOptions = new RequestClientOptions();
```
[the remark this answers is not in the document]
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/SdkRequest.java]: +1. We could call it something like `RequestOverrideConfig` to mimic the `ClientOverrideConfig` in the client builders.

**Comment by shorea** [comment]
I think that's more or less what I mean but that might violate the goal of reserving as few names as possible. I was thinking we'd put all request related config (and all response related metadata) into a single object and that's the thing the customer sets. We would extend that for API Gateway or AWS services to add the things they need with some stuff perhaps in the base. I like the idea of a common implementation that we can pass around and using interfaces to present different views of that, I think that gives us a lot of flexibility without duplicating too much work across the various client types.

```java
public class SdkRequest<ConfigT> {
    private final ConfigT sdkRequestConfig...

    public ConfigT sdkRequestConfig() {
        return sdkRequestConfig;
   }
}

public class AwsRequest extends SdkRequest<AwsRequestConfig> {
}

// These would be generated per service to include optional things like API key/ IAM auth, etc
public class MyApigServiceRequest extends SdkRequest<MyApigServiceRequestConfig> {
}

// If we wanted common things they can go in here
public interface SdkRequestConfig {
     public ProgressListener progressListener();
}

// This would add any AWS specific options
public interface AwsRequestConfig extends SdkRequestConfig {
    public AwsCredentialsProvider credentialsProvider()
}

// This would be generated for each APIG client based on the features they are using. Might extend a base APIG interface if there are standard features or we could just code gen it all in
public interface MyApigServiceRequestConfig extends SdkRequestConfig {
    public String apiKey();
}

// Common implementation with all options. May need to subclass this in each client type to mix in the appropriate interface, not sure how exactly that would look
public class DefaultSdkRequestConfig {

}

**Comment by millems** [comment]
Sounds good! We should do the same hierarchy approach for ClientOverrideConfig as we do for RequestConfig.

I don't think we need to parametrize `SdkRequest` with the type of request config, since we can return a more specific type from a child class. That'll keep our request objects simpler.

```java
public class SdkRequest {
    public SdkRequestOverrideConfig requestOverrideConfig();
}

public class AwsRequest extends SdkRequest {
    @Override
    public AwsRequestOverrideConfig requestOverrideConfig();
}
```

We'll probably need to parametrize the base builder, but that's used a lot less.

**Comment by shorea** [comment]
+1 to override and subtype.

**Comment by dagnir** [comment]
Updated the PR per feedback from @shorea and @millems:

* all request configs are now confined to a single object under `SdkRequest` to avoid name clashes
* generated per-service base request/response classes and have service request/response subclass those

I actually opted to parameterize the request config override because I couldn't find a nice way for it to work on the builders; i.e. the setter for the derived type won't override the setter for the base class.

File: core/src/main/java/software/amazon/awssdk/SdkRequest.java [hunk 2]
```diff
@@ -15,12 +15,63 @@
...
+
+    private final RequestOverrideConfigT requestOverrideConfig;
+
+    protected SdkRequest(B builder) {
+        requestOverrideConfig = builder.requestOverrideConfig();
+    }
+
+    public Optional<RequestOverrideConfigT> requestOverrideConfig() {
+        return Optional.ofNullable(requestOverrideConfig);
+    }
```
Comment (millems) [inline core/src/main/java/software/amazon/awssdk/SdkRequest.java]: Could this be Optional<? extends RequestOverrideConfig> (or whatever the parent class is) here and Optional<Concrete> in the children, so that we don't have to parametrize the base with the type?

File: core/src/main/java/software/amazon/awssdk/SdkRequest.java [hunk 3]
```diff
@@ -15,12 +15,63 @@
...
 /**
  * The base class for all SDK requests.
  *
- * TODO: SDK-specific options on the {@link AmazonWebServiceRequest} should be migrated here as part of the base-model refactor.
- *
  * @see SdkResponse
  */
-public abstract class SdkRequest {
+public abstract class SdkRequest<B extends SdkRequest.Builder<B, R, RequestOverrideConfigT>,
+        R extends SdkRequest<B, R, RequestOverrideConfigT>,
```
Comment (millems) [inline core/src/main/java/software/amazon/awssdk/SdkRequest.java]: Do we ever think we'll need to convert a non-concrete request back to a builder and then back to the request type? It would be nice if we didn't need these parameters.

We could also just remove the parent interfaces, define the methods as an abstract `Builder toBuilder()` on the request and `SdkRequest build()` on the builder and let the child classes implement the interfaces?

I figure we'll be using SdkRequest a lot, and SdkRequest<?, ?> is pretty cumbersome. We pretty much never care about the parameters. The only impact would be that we couldn't use the default methods that are defined on the interface for the SdkRequest.

File: core/src/main/java/software/amazon/awssdk/SdkRequestOverrideConfig.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        this.customQueryParameters = builder.getCustomQueryParameters();
+    }
+
+    public Optional<ProgressListener> progressListener() {
+        return Optional.ofNullable(progressListener);
+    }
+
+    public Optional<Map<String, String>> customHeaders() {
+        return Optional.ofNullable(customHeaders);
+    }
```
Comment (millems) [inline core/src/main/java/software/amazon/awssdk/SdkRequestOverrideConfig.java]: In the HTTP request/response we have the default be an empty Map<String, List<String>> instead of optional. Should we do that here too? (Same for query params)

Additional nitpick: maybe additionalHeaders and additionalQueryParameters? Just to make it clearer we're going to merge these in. We should make sure the name matches whatever is in the client configuration.

File: core/src/main/java/software/amazon/awssdk/SdkRequestOverrideConfig.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+            C extends SdkRequestOverrideConfig<B, C>> extends CopyableBuilder<B, SdkRequestOverrideConfig<B, C>> {
+        ProgressListener getProgressListener();
+
+        B progressListener(ProgressListener progressListener);
+
+        Map<String, String> getCustomHeaders();
+
+        B customHeaders(Map<String, String> customHeaders);
+
+        Map<String, List<String>> getCustomQueryParameters();
```
Comment (millems) [inline core/src/main/java/software/amazon/awssdk/SdkRequestOverrideConfig.java]: Nitpick: customQueryParameters()?

File: core/src/main/java/software/amazon/awssdk/SdkRequestOverrideConfig.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+        @Override
+        public ProgressListener getProgressListener() {
+            return progressListener;
+        }
+
+        @Override
+        public B progressListener(ProgressListener progressListener) {
+            this.progressListener = progressListener;
+            return concrete.cast(this);
```
Comment (millems) [inline core/src/main/java/software/amazon/awssdk/SdkRequestOverrideConfig.java]: We usually just do an unchecked cast with the other builders. Should we do this approach there, or that approach here?

File: core/src/main/java/software/amazon/awssdk/SdkRequest.java (hunk 3 above)
Comment (dagnir) [reply core/src/main/java/software/amazon/awssdk/SdkRequest.java]: > We could also just remove the parent interfaces,

You mean `CopyableBuilder`/`ToCopyableBuilder`? I'd be okay with that if there's no objections. I just wanted to be consistent with the rest of the codebase since we use them pretty extensively.

File: core/src/main/java/software/amazon/awssdk/SdkRequest.java (hunk 2 above)
Comment (dagnir) [reply core/src/main/java/software/amazon/awssdk/SdkRequest.java]: +1 sounds good.

File: core/src/main/java/software/amazon/awssdk/SdkRequestOverrideConfig.java (hunk 4 above)
Comment (dagnir) [reply core/src/main/java/software/amazon/awssdk/SdkRequestOverrideConfig.java]: +1 i'll make both changes.

File: core/src/main/java/software/amazon/awssdk/SdkRequestOverrideConfig.java (hunk 6 above)
Comment (dagnir) [reply core/src/main/java/software/amazon/awssdk/SdkRequestOverrideConfig.java]: Don't really have a preference. The only reason I chose to go with using this method was to avoid need to suppress the unchecked warning

File: core/src/main/java/software/amazon/awssdk/SdkRequest.java (hunk 3 above)
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/SdkRequest.java]: I agree, the loss of consistency is annoying. It's definitely just a personal preference thing. If people would prefer the consistency, I'm cool with that.

I'm a bit leery of having generics in places that we use the base class a lot (in this case, more than the child classes), especially if we are pretty much only ever going to specify the parameters as wildcards.

File: core/src/main/java/software/amazon/awssdk/SdkRequestOverrideConfig.java (hunk 6 above)
Comment (millems) [reply core/src/main/java/software/amazon/awssdk/SdkRequestOverrideConfig.java]: I don't really have a preference either.

**Comment by dagnir** [comment]
So I've come up on an a few issues when trying to remove the type parameters from `SdkRequest` and the derived classes. I've discussed these issues with @millems and want to get the everyone else's thoughts on this.

## Problems
### Going from a request to its builder doesn't work without a cast
### Overriding `getConfig` from the derived request class doesn't work without a cast

Suppose you have:

```java
class Base {
    private Config cfg;

    Base(Builder b) {
        this.cfg = b.cfg;
    }

    Optional<? extends Config> getConfig() {
        return Optional.ofNullable(cfg);
    }

    static class Builder<B extends Builder<B, C>, C extends Config> {
        private C cfg;

        Builder(Base b) {
            cfg = b.cfg; // Error, b.cfg is type Config, not C; do an unchecked cast?
        }

        B setConfig(C cfg) {
            this.cfg = cfg;
            return (B) this;
        }
    }
}

class Derived extends Base {
    Derived(Builder b) {
        super(b);
    }

    public Optional<Config2> getConfig() {
        return ?? // cast super.getConfig() ?
    }

    static class Builder extends Base.Builder<Builder, Config2> {
        Builder(Base b) {
            super(b);
        }
    }
}

class Config {

}

class Config2 extends Config {

}
```

`Base` stores `Config` but we parameterize its `Builder` so that uses can set the derived class in `setConfig`. This works fine going from `Builder` -> `Base` but going the other way round doesn't work without downcasting.

`Derived` should return `Optional<Config2>` but again not possible without a downcast.

The simple solution would be to store a second reference to `Config2` in `Derived`, and do away with the `Builder` for `Base`, and have it take `Config` as a constructor parameter. The big downside to this approach is that since `Base` no longer has a builder, there's no way to change the config if all you have is a reference to `Base`.

```java
    abstract class Base {
        private Config cfg;
 
        Base(Config cfg) {
            this.cfg = cfg;
        }
 
        Optional<? extends Config> getConfig() {
            return Optional.ofNullable(cfg);
        }
    }
 
    class Derived extends Base {
        private Config2 cfg2;
 
        Derived(DerivedBuilder b) {
            super(b.cfg2);
            cfg2 = b.cfg2;
        }
 
        public Optional<Config2> getConfig() {
            return Optional.ofNullable(cfg2);
        }
 
        static final class DerivedBuilder {
            private Config2 cfg2;
 
            DerivedBuilder(Derived b) {
                this.cfg2 = b.cfg2;
            }

            DerivedBuilder setConfig(Config2 cfg2) {
                this.cfg2 = cfg2;
                return this;
            }
        }
    }
```
### Things get complicated if you want `config` to be modifyable, so it implements `ToCopyableBuilder`. You have to carry around lots of type info:

In order to do this by implementing `ToCopyableBuilder`/`CopyableBuilder`, you have to carry around even more type information:

```java

class SdkRequestConfig<BuilderT extends Builder<BuilderT, ConfigT>,
    ConfigT extends SdkRequestConfig<BuilderT, ConfigT>> implements ToCopyableBuilder<BuilderT, ConfigT> {

    interface Builder<BuilderT extends Builder<BuilderT, ConfigT>,
        ConfigT extends SdkRequestConfig<BuilderT, ConfigT>> implements CopyableBuilder<BuilderT, ConfigT> {
            // Configs common to all requests
    }
}

class MyRequest<BuilderT extends Builder<BuilderT, RequestT, ConfigBuilderT, ConfigT>,
    RequestT extends MyRequest<BuilderT, RequestT, ConfigBuilderT, ConfigT>,
    ConfigBuilderT extends SdkRequestConfig.Builder<ConfigBuilderT, ConfigT>,
    ConfigT extends SdkRequestConfig<ConfigBuilderT, ConfigT>> implements ToCopyableBuilder<BuilderT, RequestT> {

    interface Builder<BuilderT extends Builder<BuilderT, RequestT, ConfigBuilderT, ConfigT>,
    RequestT extends MyRequest<BuilderT, RequestT, ConfigBuilderT, ConfigT>,
    ConfigBuilderT extends SdkRequestConfig.Builder<ConfigBuilderT, ConfigT>,
    ConfigT extends SdkRequestConfig<ConfigBuilderT, ConfigT>> implements CopyableBuilder<BuilderT, RequestT> {

        BuilderT requestConfig(RequestConfigT requestConfig);
        RequestConfigT requestConfig();
    }    
}
```

### Sharing a common config accross different service clients is cumbersome if either of them have service-specific request configs.

The DynamoDB client might have a DDB specific request config, `class DynamoDbRequestConfig extend SdkRequestConfig` so it's `Builder` would have `setConfig(DynamoDbRequestConfig cfg)`. However, if a customer just had a instance of `SdkRequestConfig` that they used for requests to a second service, they couldn't use it on the DDB requests without copying over to `DynamodbRequestConfig` first.

## Proposed solutions

### Don't contain the configurations in a single object
This is how the request/responses originally looked in the PR, with each configuration being a top-level member of the request object. The concern with this approach was that if we're not careful, we could potentially clash with names defined by the service model. However, if we prefix all configuration members with `sdkRequest*`, i.e. `sdkRequestTimeOut`, `sdkRequestCredentialsProvider`, it's probaby not likely.

### Don't have a class hierarchy for the override configurations
Instead of sharing a common base `SdkRequestConfig` and having other classes extend that, just make them completely separate classes:

```java
final class SdkRequestConfig {
}

final class AwsRequestConfig {
}

final class DynamoDbRequestConfig {
}

class SdkRequest {
    public SdkRequestConfig sdkRequestConfig() {
        ...
    }

    static class Builder {
        public Builder sdkRequestConfig(SdkRequest config) {
            ....
        }
    }
}

class AwsRequest extends SdkRequest {
    public AwsRequestConfig awsRequestConfig() {
        ...
    }

    static class Builder extends SdkRequest.Builder {
        public Builder awsRequestConfig(AwsRequestConfig config) {
            ...
        }
    }
}

class DynamoDbRequest extends AwsRequest {
    public DynamoDbRequestConfig DynamoDbRequestConfig() {
        ...
    }

    static class Builder extends AwsRequest.Builder {
        public Builder dynamoDbRequestConfig(DynamoDbRequestconfig config) {
            ...
        }
    }
}

// and later
PutItemRequest putItem = PutItemRequest.builder()
                                       .skdRequestConfig(...)
                                       .awsRequestConfig(...)
                                       .dynamoDbRequestConfig(...)

```

The downside to this is that might be difficult to discover which config objects contain what settings.

### Don't bother trying to return derived types in the config getters

Basically, don't even overload the getter with a more derived type, and just check at runtime

```java
class SdkRequest {
    final SdkRequestConfig sdkRequestConfig();

    static class Builder {
        final Builder sdkRequestConfig(SdkRequestConfig config) {
            ...
        }
    }
}

// and later
if (request.sdkRequestConfig() instanceof AwsRequestConfig) {
    ...
}

if (request.sdkRequestConfig() instanceof DynamoDbRequestConfig) {
    ...
}
```

Personally, I like both of the first two approaches.

**Comment by millems** [comment]
+1 for "Don't contain the configurations in a single object"

File: core/src/main/java/software/amazon/awssdk/SdkRequest.java (hunk 3 above)
Comment (bpholt) [reply core/src/main/java/software/amazon/awssdk/SdkRequest.java]: > Do we ever think we'll need to convert a non-concrete request back to a builder and then back to the request type?

As a maintainer of a Scala library that uses [generic programming to abstract pagination](https://github.com/Dwolla/scala-aws-utils/blob/fs2/fs2AwsUtils/src/main/scala/com/dwolla/awssdk/PaginatedAwsClient.scala) (creating [fs2 `Stream`](https://github.com/functional-streams-for-scala/fs2) for a given list of data (e.g. CloudFormation stacks or ECS services)), I would say yes, it is helpful to be able to convert a generic request back to a builder of the appropriate type in order to create a new request with modifications.

With the 1.11 SDK, I'm using structural types to identify what [`AmazonWebServiceRequest`](https://github.com/Dwolla/scala-aws-utils/blob/fs2/core/src/main/scala/com/dwolla/awssdk/utils/package.scala#L12) and [`AmazonWebServiceResponse`](https://github.com/Dwolla/scala-aws-utils/blob/fs2/core/src/main/scala/com/dwolla/awssdk/utils/package.scala#L14) types are paginated. Given a base request of the right type and a reference to an [async request function](https://github.com/Dwolla/scala-aws-utils/blob/fs2/core/src/main/scala/com/dwolla/awssdk/utils/package.scala#L15), it can fetch all the pages for the caller and provide the data as an fs2 stream. This relies on being able to construct a new request object for each page it needs to request, in order to set the pagination token for the next request.

Right now I'm hacking around it by passing around a request factory function `() ⇒ PaginatedRequest`, where 

```scala
type PaginatedRequest = AmazonWebServiceRequest {def setNextToken(s: String): Unit}
```

(that's a nullary function that returns an instance of type `PaginatedRequest`, which is a subtype of `AmazonWebServiceRequest` with method `setNextToken` having the right signature), which is called to create each request, and then modified with the next pagination token.

I'm still trying to wrap my head around the structure of the new types, but I think it would definitely be useful to be able to create new instances of a given type rather than have to pass around the factory function I'm using now.

I'm not sure what the right level of detail is here. I would just ask that as you make design decisions of this type, you keep this type of thing in mind. Let me know how I can be more helpful.

(FWIW, it would also be nice to have an interface for pagination so I don't have to rely on structural typing (which uses reflection at runtime, with the corresponding runtime penalties) to deal with the pagination tokens.)

Comment (millems) [reply core/src/main/java/software/amazon/awssdk/SdkRequest.java]: That's great info, thanks! I think it definitely makes sense that people should be able to do this then, at least at the `AwsRequest` level, which is the level most people care about.

> FWIW, it would also be nice to have an interface for pagination so I don't have to rely on structural typing (which uses reflection at runtime, with the corresponding runtime penalties) to deal with the pagination tokens.

Because you've given great feedback, I feel like I want to ask for more. Would you mind giving yours on our automatic depagination pull request to see if it fits your needs? https://github.com/aws/aws-sdk-java-v2/pull/207

**Comment by dagnir** [comment]
Closing in favor of #289.
